### PR TITLE
core: expose engine via weak_ptr

### PR DESCRIPTION
**Pull request by goaway**
Description: Access Engine via weak_ptr so that threads can extend its lifetime for the durations of their calls to it. Despite the fact that all incoming calls are ultimately dispatched to the underlying Envoy main thread, there exists a window where the thread must leverage Engine state (e.g. to access the dispatcher) that can otherwise race with the Engine lifecycle. While it's probably possible to cover these edge cases at the platform layer, this provides a common solution.
Risk Level: Moderate
Testing: local and CI

Signed-off-by: Mike Schore <mike.schore@gmail.com>

### CODE REVIEW

**Comment by goaway** [comment]
@junr03 So I’m kinda going back on what we talked about Friday - I do think leaking forever would fix our problems for now, but I still think a weak pointer is probably just necessary long-term. Because the engine’s lifecycle will(?) ultimately be tied to the platform wrapper’s lifecycle, we need to be very careful about the timing of incoming calls coming from other (arbitrary) threads. I think we will always want to manage the engine’s state via an exposed weak_ptr that we use to capture shared_ptrs for the duration of calls - to prevent destruction starting while _any_ thread might be touching Engine in its stack.

To make this non-singleton, we convert the internal shard_ptr to one the platform wrapper holds (and eventually resets).

File: library/common/main_interface.cc [hunk 1]
```diff
@@ -56,6 +71,8 @@ envoy_status_t run_engine(envoy_engine_t, envoy_engine_callbacks callbacks, cons
                           const char* log_level) {
   // This will change once multiple engine support is in place.
   // https://github.com/lyft/envoy-mobile/issues/332
-  engine_ = std::make_unique<Envoy::Engine>(callbacks, config, log_level, preferred_network_);
+  static auto internal =
```
Comment (junr03) [inline library/common/main_interface.cc]: Not a big deal, but for consistency of our statics, should we declare this above as internal_engine_ or strong_engine_, or better_name_?

Comment (goaway) [reply library/common/main_interface.cc]: Definitely we can change the name, but I did think it made more sense, at present, to encapsulate its declaration in this function call:
- nothing else should ever touch it, and this limits its accessible scope, and
- its lifecycle is more deterministic here.

Comment (junr03) [reply library/common/main_interface.cc]: Thanks for the explanation. I think the first point is sufficient on its own to leave it here.

**Comment by goaway** [comment]
Thanks for the review :)

**Pull request merged**